### DISCUSSION TITLE
Feature request: Allow numbers in friendly names

### DISCUSSION THREAD

**Discussion by stephan-henningsen** [Koenkk/zigbee2mqtt#11352] [General]
Hi,

I'd like to be able to name my devices with a number, e.g:
  - `zigbee2mqtt/power-switch/1`
  - `zigbee2mqtt/power-switch/2`
  - `zigbee2mqtt/push-button/1`
  - `zigbee2mqtt/push-button/2`

I see that [others](https://github.com/Koenkk/zigbee2mqtt/issues/2645) do [too](https://github.com/Koenkk/zigbee2mqtt/issues/2309).

Because first of all, it makes sense.  I know mqtt isn't REST, but it's nice to be able to reuse meaningful identifiers like this in this domain.

Secondly, I would like to use mqtt topic filter (in nodered) like `zigbee2mqtt/power-switch/+` to process all power switches.  Mqtt doesn't support filtering part of a topic leves, ie. this is _not_ possible: `zigbee2mqtt/power-switch-*` to subscribe to `zigbee2mqtt/power-switch-1`, `zigbee2mqtt/power-switch-2`, etc.

I see that the numbers are used for endpoints.  But are these ever addressed by the clients of zigbee2mqtt directly?  _**If**_ these endpoints are used strictly internally to zigbee2mqtt, could it be possible to expose them as such, and not interfere with the public side of the API?  Perhaps use `zigbee2mqtt/__endpoint/1` or maybe `zigbee2mqtt/__internal/endpoint/1` if more internal stuff exists?

Ideally, this change could be rolled out and not affect any users. Less ideally but still better, it could be rolled out as a new major version, thereby signalling a breaking change in API.

What do you say? Could this work? @Koenkk?

Yours,
Stephan

**Comment by Koenkk** [comment]
> I see that the numbers are used for endpoints. But are these ever addressed by the clients of zigbee2mqtt directly?

Since this is currently possible, yes we have to assume it is (and this looks valid to me). I currently don't see a reason to change this. 

As a workaround you could use something like `zigbee2mqtt/power-switch/n1`

**Comment by stephan-henningsen** [comment]
Hi again.

I've found solution that doesn't break anything:

Add a new configuration option.  It should be disabled by default so that behavior isn't changed and everything keeps working as currently.  Specifically, no breaking API change is introduced and `zigbee2mqtt/power-switch/1` refers to the endpoint 1 of the zigbee device named `power-switch`.

When the option is _actively_ enabled, it gives the [REST](https://restfulapi.net/rest-api-design-tutorial-with-example/)-like mqtt topic structure that many of us would like. Zigbee2mqtt does this by injecting the `endpoint` topic level.  This would allow devices to have numbers `zigbee2mqtt/power-switch/123` and have endpoints available at `zigbee2mqtt/power-switch/123/endpoint/1`.

So `configuraiton.yaml` could have this `device_options/inject_endpoint_topic` option added:
```
device_options:
  inject_endpoint_topic: true
```

Or similar to `mqtt/base_topic` a custom endpoint topic could be set to allow more flexibility:
```
device_options:
  endpoint_topic: endpoint
```

Programmatically, I'm guessing you could set an `endpoint_topic` variable to `""` and concat `settings.get().device_options.endpoint_topic` everywhere endpoints are referenced, preferably high up where [mqtt](https://github.com/Koenkk/zigbee2mqtt/blob/master/lib/mqtt.ts) is handled so the models and other processing isn't modified.  Then let the configuration parser override it, possible add a trailing slash.   I'd like to make the pull request myself, only I'm not very much into js/ts =\.

Anyway, I think this is a viable solution that works to everyone's advantage, don't you?

**Reply by Koenkk** [reply]
This would be an option indeed, however I only want to add this when more people are interested in such feature.
